### PR TITLE
Draft point cloud STAC extension

### PR DESCRIPTION
**Pull request by hobu**
A draft point cloud STAC extension based on [PDAL info](https://pdal.io/apps/info.html) output for geospatial point clouds.

### CODE REVIEW

File: extensions/pointcloud/example-alaska.json [hunk 1]
```diff
@@ -0,0 +1,261 @@
+{
+  "assets": {},
+  "bbox": {
```
Comment (m-mohr) [inline extensions/pointcloud/example-alaska.json]: GeoJSON bboxes aren't objects but arrays, I think.

File: extensions/pointcloud/example-alaska.json [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+      ]
+    ],
+    "type": "MultiPolygon"
+  },
+  "id": "AK_MatanuskaSusitna-Lot2_2011_000227.laz",
+  "properties": {
+    "c:description": "USGS 3DEP LiDAR",
+    "c:id": "MODIFICATION                   ",
+    "datetime": "2013-05-14T00:00:00",
+    "license": "LICENSE",
```
Comment (m-mohr) [inline extensions/pointcloud/example-alaska.json]: Property name should be `item:license` (see Single Item extension)

File: extensions/pointcloud/example-alaska.json [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            61.76682620765713
+          ]
+        ]
+      ]
+    ],
+    "type": "MultiPolygon"
+  },
+  "id": "AK_MatanuskaSusitna-Lot2_2011_000227.laz",
+  "properties": {
+    "c:description": "USGS 3DEP LiDAR",
```
Comment (m-mohr) [inline extensions/pointcloud/example-alaska.json]: c:id and c:description are given, but no collection is linked to. Would remove both properties.

File: extensions/pointcloud/example-alaska.json [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        "stddev": 10532949.21,
+        "variance": 110943019000000.0
+      }
+    ],
+    "pc:type": "lidar",
+    "pc:units": {
+      "horizontal": "US survey foot",
+      "vertical": "US survey foot"
+    },
+    "provider": "USGS"
```
Comment (m-mohr) [inline extensions/pointcloud/example-alaska.json]: Property name should be `item:provider` (see Single Item extension)

File: extensions/pointcloud/example-alaska.json [hunk 5]
```diff
@@ -0,0 +1,261 @@
+{
```
Comment (m-mohr) [inline extensions/pointcloud/example-alaska.json]: The example is missing the required `links` property.

File: extensions/pointcloud/pdal-to-stac.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
```
Comment (m-mohr) [inline extensions/pointcloud/pdal-to-stac.py]: Script should be adopted to the proposed changes for the example file.

File: extensions/pointcloud/README.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+## Item Fields
+
+| Field Name       | Type                 | Description                                                                                                                                                                                                                                  |
+| ---------------- | -------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| pc:count | integer | The number of points in the item
+| pc:density| float | Number of points per square unit area
+| pc:encoding| string | Content encoding or format of the data
+| pc:schema| array | A sequential array of items that define the dimensions and their types
+| pc:statistics| array | A sequential array of items mapping to `pc:schema` defines per-channel statistics
```
Comment (m-mohr) [inline extensions/pointcloud/README.md]: Type should be `[Statistics Object]` (or similar) and the Statistics Object should be specified, see other extensions (EO, SAR) for examples.

File: extensions/pointcloud/README.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+**Extension [Maturity Classification](../README.md#extension-maturity): Pilot**
+
+This document explains the fields of the STAC Point Cloud Extension to a STAC Catalog,
+which allows STAC to more fully describe point cloud datasets. The point clouds can
+come from either active or passive sensors, and data is frequently acquired using
+tools such as LiDAR or coincidence-matched imagery.
+
+- [Example](example-alaska.json)
+- [JSON Schema](schema.json)
```
Comment (m-mohr) [inline extensions/pointcloud/README.md]: There is a link to a schema.json, but this file doesn't exist.

File: extensions/pointcloud/README.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+## Item Fields
+
+| Field Name       | Type                 | Description                                                                                                                                                                                                                                  |
+| ---------------- | -------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| pc:count | integer | The number of points in the item
+| pc:density| float | Number of points per square unit area
+| pc:encoding| string | Content encoding or format of the data
+| pc:schema| array | A sequential array of items that define the dimensions and their types
+| pc:statistics| array | A sequential array of items mapping to `pc:schema` defines per-channel statistics
+| pc:type| enumeration | Phenomenology type for the point cloud. Valid values include `lidar`, `eopc`, `radar`, `sonar`
```
Comment (m-mohr) [inline extensions/pointcloud/README.md]: JSON doesn't have a data type "enumeration". Should be simply "string". Valid values are listed in the description anyway. Are the valid values all possible values or can there be others? In the second case it wouldn't be an enumeration anyway.

File: extensions/pointcloud/README.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
+# STAC Point Cloud Extension Spec (`pc`)
+
+**Extension [Maturity Classification](../README.md#extension-maturity): Pilot**
```
[the remark this answers is not in the document]
Comment (cholmes) [reply extensions/pointcloud/README.md]: D'oh - this one is my bad. Added this in, meant to do proposal and then just forgot to edit my copy and paste. Will fix.

File: extensions/pointcloud/README.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+- [JSON Schema](schema.json)
+
+## Item Fields
+
+| Field Name       | Type                 | Description                                                                                                                                                                                                                                  |
+| ---------------- | -------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| pc:count | integer | The number of points in the item
+| pc:density| float | Number of points per square unit area
+| pc:encoding| string | Content encoding or format of the data
+| pc:schema| array | A sequential array of items that define the dimensions and their types
```
[the remark this answers is not in the document]
Comment (m-mohr) [reply extensions/pointcloud/README.md]: I just added a boilerplate for you to fill the missing descriptions. Hope that makes it easier to understand what I was missing.

File: extensions/pointcloud/README.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+- [Example](example-alaska.json)
+- [JSON Schema](schema.json)
+
+## Item Fields
+
+| Field Name       | Type                 | Description                                                                                                                                                                                                                                  |
+| ---------------- | -------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| pc:count | integer | The number of points in the item
+| pc:density| float | Number of points per square unit area
```
[the remark this answers is not in the document]
Comment (m-mohr) [reply extensions/pointcloud/README.md]: Fixed it for your convenience.

File: extensions/pointcloud/README.md (hunk 7 above)
Comment (m-mohr) [reply extensions/pointcloud/README.md]: I just added a boilerplate for you to fill the missing descriptions. Hope that makes it easier to understand what I was missing.

File: extensions/pointcloud/README.md (hunk 8 above)
Comment (m-mohr) [reply extensions/pointcloud/README.md]: Fixed it for your convenience.

File: extensions/pointcloud/example-alaska.json [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+        ]
+      ]
+    ],
+    "type": "MultiPolygon"
+  },
+  "id": "AK_MatanuskaSusitna-Lot2_2011_000227.laz",
+  "properties": {
+    "c:description": "USGS 3DEP LiDAR",
+    "c:id": "MODIFICATION                   ",
+    "datetime": "2013-05-14T00:00:00",
```
Comment (m-mohr) [inline extensions/pointcloud/example-alaska.json]: Timezone (e.g. `Z`) is missing, this would not pass validation.

Comment (hobu) [reply extensions/pointcloud/example-alaska.json]: that datetime came from the file itself, which is often missing such information. Presumably the script creating the STAC item from the `pdal info` output would need to stick one in there.

File: extensions/pointcloud/README.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+### Schema Object
+
+A sequential array of items that define the dimensions and their types. All fields are
+required.
+
+| Field Name | Type    | Description |
+| ---------- | ------- | -------------------------- |
+| name       | string  | The name of the dimension. |
+| size       | integer | The size of the dimension in bytes. Whole bytes only are supported.|
+| type       | string  | Dimension type. Valid values include `floating`, `unsigned`, and `signed`|
```
Comment (m-mohr) [inline extensions/pointcloud/README.md]: As above, it is potentially a bit easier to understand as follows: "Possible valid values might include `floating`, `unsigned`, and `signed`." (Or are these values exclusive?)

Comment (hobu) [reply extensions/pointcloud/README.md]: They're exclusive, and a dimension can only have a single type of one of those three enumerations.

Comment (m-mohr) [reply extensions/pointcloud/README.md]: Great, thanks. Resolving this.

**Comment by m-mohr** [comment]
By the way, should any of the fields be required or are these all optional (which they are currently)?

**Comment by hobu** [comment]
Some of them should be required. I'll go through and denote.

**Comment by m-mohr** [comment]
> Some of them should be required. I'll go through and denote.

We usually do be prefixing the descriptions with the following markdown `**REQUIRED.** ` - Thanks.

**Pull request merged**